### ISSUE TITLE
Currency in register

### ISSUE THREAD

**Issue by dvidic** [opensourcepos/opensourcepos#1988]
### Installation information

- OSPOS version is: 3.2.0

### New Feature

I have a kind request to good people developing OSPOS; 
possibility to make receipt/invoice in two (or more) currencies. 
For example, I have all item stored in my currency (HRK), and because we are not in eurozone, sometime I have to make an invoice in euro. It would be great to make a possibility to issues a invoice in euros/other currency. I guess that only currency conversion number should be entered (it could be manually) and OSPOS should calculate all values in euros. 

I don't know is is necessary to make a double register for that or is it possible to make in in current one.  Invoice in euro is different to one in my home currency only in values and it needs just one sentence below the calculated valus (Invoice is without VAT...)

Please, can something like this be done? Europe has many countries with this situation, doing bussines in two currencies. 

Thank you!!!

**Comment by mckaygerhard** [comment]
umm ihave an idea, due it's not so easy.. 

a popup calculator! i mean wheantime due this feature its not so easy event put two currentcy's, put a popup calculator, to convert beetween currentcy's!

this can solve until some more complex are in develop!

**Comment by SteveIreland** [comment]
Interesting request.  So you would only need the alternate currency for the purpose of the sale?  On the database could the amounts continue to be stored in your default currency and the alternate currency be for display in the register and for document printing?  Or would it need to be stored in the database with the selected currency?  The second option will present all sorts of issues when it comes to reporting.

What triggers when an alternate currency has to be used?

**Comment by mckaygerhard** [comment]
@SteveIreland i choose the first option. due this are more easyle to implement.. alternate currentcy can be show as option, and default currentcy could be a converted amount of the alternate (and then stored in DB..)

for the reports, can show the alternate currency only if are present with a sub-select query.. so then will not affect current selects of reports..

**Comment by dvidic** [comment]
@SteveIreland  alternate currency only for sale/document purposes. Items remain stored with default currency. Only addtition is that this kind of invoice (in alternate currency) must have disclaimer below total price (note that invoice is VAT free). 
As I said above, currency conversion could be done by manual entering multiplicator,.for.example 1 euro is 7.40HRK, so I would enter 7.40 in the field, if avalaible. I don't know could it be done by auto syncronization of conversion values, but that's not that important.

**Comment by mckaygerhard** [comment]
@dvidic you enhanced my idea, the multiplicator can be by a popup select, and user must5 enter the multiplication factor (due the currently has changes beetween countryes respect economic flutues)

**Comment by dvidic** [comment]
@mckaygerhard Yes, exchange rate varies, so manual entering or pulling from some web service is an option.

**Comment by mckaygerhard** [comment]
pulling from som web services are not viable, by example in Venezuela there's no service that put a official price respect currentcy.. i think its the mayor example! some other countries applied perfectly to this example, such Panama, Dominicana, India...

**Comment by nerv55** [comment]
@dvidic What is the regulation there?, for example you official currency is HRK but you can print invoce in Euro. is that permited in your country?.
Here [Dominica Republic] you can show prices in Dollars but all the invoices must be printed in Pesos [Oficial Currency] and if you accept paiment in dollars you must print the exchange rate used in the transaction.

**Comment by dvidic** [comment]
Croatia is part of EU, but has not adopted euro, probably around 2022. We can issue invoice in our currency (kuna), but we can also issue a invoice in euro (without VAT) if customer needs it.

**Comment by mckaygerhard** [comment]
if euro stil are not adopted, its like venezuela transition from Bs to BsS only by few months are double currentcy.. so i doub that ospos developers wil made a feature only for a few months of transition

**Comment by dvidic** [comment]
No, as I state above, it's not for a few months. It's at least 5 years. And also, this feature is useful for export invoices...

**Comment by dvidic** [comment]
@SteveIreland can You do something about this enhancemet, please..

**Comment by SteveIreland** [comment]
Sure, but only if you promise to buy me a shot of whiskey at the first International OSPOS Conference.

**Comment by WebShells** [comment]
@SteveIreland we should do members meeting, let's discuss it seriously :x

**Comment by dvidic** [comment]
@SteveIreland  It can be done ;)
Please, when you get the time, try to implement this enhancement.

**Comment by SteveIreland** [comment]
@WebShells I'm serious about my whiskey.

**Comment by WebShells** [comment]
You deserve a bottle 🍾 :p Consider JD Bottle in the first OSPOS conference 😎

**Comment by SteveIreland** [comment]
I looked at this a bit today, and obviously it's loaded with mine fields.  The first obvious one was we have a LOT of support for specialized locale configuration.  If we add support for generating an invoice in a different currency then it seems like we would need a lot of that locale configuration replicated for the alternate currency so that the sales documents will be printed correctly. I am thinking that the following configuration would be require:

1. alternate currency
2. alternate currency current conversion factor
3. alternate currency symbol
4. Alternate currency thousands separator

I'm leaning toward adding a check box on the register that says Use alternate currency.  If it is checked then it will pop up the currency code and the current conversion factor that can be overridden.  If it is overridden then it will update the configuration and will be used for the next sale that requires an alternate currency.

Am I wrong?   Is there anything else that I'm missing?

**Comment by mckaygerhard** [comment]
well, "alternate currentcy thousands separator" might implicts "decimal separatos" too and i think could use the current.. and may implict many other math calculations.

i dont agree with have generated in "other" directly, think its beter alternate table data.. and only few calculations on the fly..  so all the point that implicts will be charged by the locale conversion..

**Comment by daN4cat** [comment]
Currency conversion factor could change daily. How are you going to keep it up to date?
Not sure this should really make the OSPOS master as it sounds a corner case.

**Comment by mckaygerhard** [comment]
umm but how about  ask to current conversion fator on each time, and also show the last entered conversion as @SteveIreland said?

**Comment by SteveIreland** [comment]
Work?  How can it be "work" when it is a passion?  Fishing requires more effort.  Everyone should be so lucky as to solve puzzles for a living.

**Comment by dvidic** [comment]
People who need this will enter factor. 
This is not a corner case, 19 countries in EU use euro, 9 not. If you want to sell product to buyer from eurozone, you must issue a invoice in euro (without VAT). With current conditions in OSPOS, we are unable to issue such invoice.

**Comment by dvidic** [comment]
@SteveIreland pls implement this and leave me paypal adress so I can buy you a shot of whiskey ;)

**Comment by WebShells** [comment]
@dvidic @SteveIreland needs a bottle 🍾 🥃... Am really planning for a conference for the organization members.. @jekkos @daN4cat @SteveIreland @odeia @objecttothis? You agree? And whoever would like to join us 😎

**Comment by dvidic** [comment]
@SteveIreland is there any chance of implementation?

**Comment by SteveIreland** [comment]
There are two pull requests I would like to see merged before proceeding with this. #2027 and #1987 
I think #1987 might impact how I do some coding so I really need that installed.  #2027 includes changes to programs that I'm messing with on other issues so it would be nice to have that installed so I don't need to worry about making the merge a nightmare.

I also have the India GST changes in progress.

It gets troublesome for me when there are too many things in play at the same time - especially when one of those things might be as complex as I think this might become.

**Comment by dvidic** [comment]
@SteveIreland **thank You m8 in advance.** 
Looking forward to this feature.
As I said before, I'll be happy to buy You whiskey for all this effort.

**Comment by dvidic** [comment]
@SteveIreland did You found time to make a progress on this one, maybe?
Thank You.

**Comment by SteveIreland** [comment]
Not yet. The India GST has consumed all of my available time but it's getting close to prototype.  However, I will probably extend it a week and take a look at the item attributes pull request this weekend to see how I might take advantage of it and how it will impact my development efforts.

I expect that the attributes project will be merged before anything I submit so I'd need to rebase my work on it anyway.

However, Currency in Register is next in the queue.

**Comment by dvidic** [comment]
@SteveIreland m8, is this one next?

Thank You!

**Comment by SteveIreland** [comment]
This is next on the agenda but the priority continues to be the India GST project.  I need to make sure that the taxes are calculated and presented correctly in all possible permutations ... and there are a LOT of permutations.

**Comment by dvidic** [comment]
@SteveIreland I sure hope you haven't forgot about this one ;)
I see India GST project is finished...
THANK YOU

**Comment by SteveIreland** [comment]
I was going to ask if you were still interested.  I'm finishing up some rework on the india-gst project that @jekkos asked for.  This is next on the list.

**Comment by dvidic** [comment]
Yes, very interested. Thanks for all your effort!

**Comment by dvidic** [comment]
@SteveIreland did You found some time for this one, m8? Any progress, maybe?
Thank you.

**Comment by SteveIreland** [comment]
Starting it now ... your patience is amazing.

**Comment by SteveIreland** [comment]
Here is what is currently being built.  Let me know if I'm not on track.

I'm only adding support for one alternate currency at a time.

At the point of sale, if alternate currency is enabled then the user has an option to print a a document in an alternate currency.  The user selects a check box to print in an alternate currency and enters the exchange rate to be used.  This has no effect on the amounts displayed in the register since it seems to me that those numbers should still be in the local currency. 

If the receipt or invoice is printed and the "Use Alternate Currency" flag is checked then all of the currency amounts will have the exchange rate printed and the number will be rendered using the alternate currency's format (as interpreted by the NumericFormat class.

The exchange rate and the alternate language code in use at the time is saved with the sale so that the receipt or invoice can be reprinted using the alternate currency and rate in play at the time of the original sale.

**Comment by dvidic** [comment]
Perfect!

**Comment by SteveIreland** [comment]
Update:  The code to print receipts and invoices in an alternate currency is complete.  I'm in the process of adding support for reprinting the documents using the alternate currency from "Daily Sales".  Since it might be necessary to reprint it because the exchange rate was entered incorrectly I thought it would be best to allow the user to correct the exchange rate and then print the document in its alternate currency.

**Comment by WebShells** [comment]
Steveeee, can you have a look on singls button pay out? 1,5,10,20,50,100 usd payment buttons that can finish sales on 1 click? Including completion?

**Comment by SteveIreland** [comment]
Pull request #2375 has been submitted to address this issue

**Comment by SteveIreland** [comment]
Documentation for this feature is found on the wiki at https://github.com/opensourcepos/opensourcepos/wiki/Alternate-Currency-Receipts-and-Invoices

**Comment by SteveIreland** [comment]
@WebShells I'm going to need to pass on the denomination buttons.  The feature I really should start working on is the purchase order management feature.

**Comment by discapacidad5** [comment]
I did the installation of this program because I loved this option of the double currency but it is searched for all the forms and it does not appear to me to establish the double currency I thought it was my installation and I went to the demo version and it does not appear either could you inform me a little more like being able to use this alternative.

**Comment by daN4cat** [comment]
@discapacidad5 I'm afraid it's not implemented.

**Comment by discapacidad5** [comment]
In the tutorials appears and appears the way how to enable it in the system there is no such option already install even the latest version in development and is not.

https://github.com/opensourcepos/opensourcepos/wiki/Alternate-Currency-Receipts-and-Invoices

**Comment by ekardian** [comment]
oooht its here the hilo, i migrated from bolivian money to US dollars manually all my little db for start in production, in a little drinking store, not wisky of coorrs xD, wold be great the currency switcher, like the wordpress currency switcher? no problem i can wait a few years more, if i was wating 7years, why not other 7 years more
